### PR TITLE
Allow audio components to be wrapped as drawables

### LINKED ISSUE #2125: Move audio components to the draw hierarchy
The audio subsystem currently operates as its own hierarchy and misses out of the benefits of existing in the same place as everything else. With this change, we will allow all audio properties to be recursively and locally adjusted, and better manage lifetimes of audio components (as they will be spawned/cleaned with containers/screens/etc.

- `AudioManager` would be at a `Game` level and manage overall volume as it currently does.
- `AdjustableAudioComponents`, `Tracks` etc. could be added anywhere in the hierarchy. We would use DI or passing of state to maintain integrity in the hierarchy (similar to how `clock` is passed down currently).
- It should be possible to move a `Track` from one location to another and have it synchronise with its new additive/multiplicative parent adjustments.

This is a relatively high priority change as we need arbitrary local parameter adjustments and transforms on these parameters. After implementation we will be able to use the same transform system we use on `Drawable`s on `AudioComponent`s too.

Quick brainstorm on how I see this working:

image

If we decide to use the DI approach, we will need to reassess the fact that DI is not performed a second time when a parent is changed.

Prerequisites:

- Possibly rename the draw hierarchy internally
- Make `Component` the base class of `Drawable`, rather than vice-versa (how it currently is for simplicity).
- Assess how clocks and audio components fit together. In many cases an audio component also acts as a clock.

### PR DESCRIPTION
**Pull request by peppy**
Focuses on getting an idea of how this will feel. This is done by moving as little logic as possible from existing classes while fully enabling use of samples in the draw hierarchy.

- Naming is temporary (Component* prefix is used for new classes).
- Tracks still need some tests.
- More tests to follow.
- Closes #2125.
- [x] Depends on #2447.
- [x] Depends on #2449.
- [x] Depends on #2451.
- [x] Depends on #2453.
- [x] Depends on #2458.
- [x] Depends on #2459.

RFC regarding the structure of this.

### CODE REVIEW

File: osu.Framework/Audio/ComponentAudioAdjustContainer.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics.Containers;
+
+namespace osu.Framework.Audio
+{
+    [Cached(typeof(IAudioAdjustment))]
+    public class ComponentAudioAdjustContainer : Container, IAudioAdjustment
```
Comment (smoogipoo) [inline osu.Framework/Audio/ComponentAudioAdjustContainer.cs]: As discussed, having all audio components be containers feels wrong. I think a better structure for this is one similar to `DelayedLoadWrapper`, that is:

```
AudioWrapper : Drawable
    ctor(Drawable wrappedDrawable)

AudioSample : AudioWrapper
    ctor() : base(null)

AudioContainer : AudioWrapper, IContainerEnumerable<T>, IContainerCollection<T>, ICollection<T>, IReadOnlyList<T>
    ctor() : base(new Container())
    /* forward all interface members to the wrapped container */
```

**Comment by UselessToucan** [comment]
I suggest to move `Drawable` classes to a separate namespace like `osu.Framework.Audio.Drawables`.

**Comment by peppy** [comment]
Oh definitely. Namespaces have not yet been considered.

**Comment by peppy** [comment]
This is ready for review 👍 .

**Pull request merged**